index.js

```diff
@@ -1,130 +1,163 @@
+  const timestamp = +new Date();
+  const mainModule = `RunElmMain${timestamp}`;
+  const templatePath = path.resolve(__dirname, 'RunElmMain.elm.template');
+  const templateWithArgsPath = path.resolve(__dirname, 'RunElmMainWithArgs.elm.template');
+  const userModulePath = path.resolve(rawUserModuleFileName);
+  const userModule = path.basename(userModulePath, '.elm');
+  const projectDir = rawProjectDir
+    ? path.resolve(rawProjectDir)
+    : path.resolve(path.dirname(userModulePath));
+  const mainModuleFilename = path.join(projectDir, `${mainModule}.elm`);
+  const outputCompiledFilename = path.join(projectDir, `run-elm-main-${timestamp}.js`);
+
+  let exitCode = 0;
+  try {
+    // ensure elm is installed
+    if (!sh.which('elm')) {
+      throw new Error('No elm global binary available. Please install with `npm install -g elm`.');
+    }
 
-  const contents = await compileToString([outputElmFilename], { yes: true });
+    // ensure --output-name is specified adequately
+    if (!outputName.match(/^[a-z_]\w*$/)) {
+      throw new Error(`Provided --output-name \`${outputName}\` is not a valid constant or function name in elm.`);
+    }
+    if (['init', 'main', 'program', 'sendOutput'].includes(outputName)) {
+      throw new Error(`It is not allowed to use \`${outputName}\` as a value for --output-name. Please rename the variablue you would like to output.`);
+    }
 
-  return writeFile(outputCompiledFilename, contents);
-}
+    // ensure user module path is adequate
+    try {
+      const userModuleFileStats = await stat(userModulePath);
+      if (!userModuleFileStats.isFile()) {
+        throw new Error();
+      }
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' does not exist`);
+    }
+    if (!userModulePath.match(/\.elm$/i, '')) {
+      throw new Error(`File '${userModulePath}' should have .elm file extension`);
+    }
 
-function runElmFile(needArgs) {
-  return new Promise((resolve) => {
-    const { worker } = require(outputCompiledFilename)[runModuleName];
-    let app;
+    // ensure project folder is adequate
+    try {
+      const projectDirStats = await stat(projectDir);
+      if (!projectDirStats.isDirectory()) {
+        throw new Error();
+      }
+    } catch (err) {
+      throw new Error(`Provided --project-dir '${rawProjectDir}' is not a directory`);
+    }
+    if (userModulePath.indexOf(`${projectDir}/`) !== 0) {
+      throw new Error(`File \`${userModulePath}\` must be located within --project-dir \`${rawProjectDir}\``);
+    }
 
-    if (needArgs) {
-      const args = process.argv.slice(3);
-      app = worker(args);
-    } else {
-      app = worker();
+    // read user module and determine what template to use
+    let userModuleContents;
+    try {
+      userModuleContents = await readFile(userModulePath, 'utf-8');
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' could not be read`);
+    }
+    const argsRegex = new RegExp(`^${outputName} +\\w+ *=`);
```

Comment:
I completely forgot to replace `/^${outputName} +\\w+ *=/` with a more general case and only discovered this after a new test failed 😆